certora/spec/KintoID/SanctionsTraits.spec

```diff
@@ -0,0 +1,397 @@
+
+    assert stateA[currentContract] == stateB[currentContract];
+}
+
+rule removeSanctionCommutativity() {
+    env e;
+    address accountA; uint16 CID_A;
+    address accountB; uint16 CID_B;
+    
+    storage initState = lastStorage;
+        removeSanction(e, accountA, CID_A) at initState;
+        removeSanction(e, accountB, CID_B);
+    storage stateA = lastStorage;
+        removeSanction(e, accountB, CID_B) at initState;
+        removeSanction(e, accountA, CID_A);
+    storage stateB = lastStorage;
+
+    assert stateA[currentContract] == stateB[currentContract];
+}
+
+rule addedSanctionCanBeRemoved(address account, uint16 CID) {
+    env e1;
+    env e2; require e2.msg.value == 0; 
+    requireInvariant hasSanctionCountIsNonZero(e1, account, CID);
+    requireInvariant hasSanctionCountIsNonZero(e2, account, CID);
+
+    bool hasRole1 = hasRole(KYC_PROVIDER_ROLE(), e1.msg.sender);
+    bool hasRole2 = hasRole(KYC_PROVIDER_ROLE(), e2.msg.sender);
+    addSanction(e1, account, CID);
+    assert hasRole1;
+
+    removeSanction@withrevert(e2, account, CID);
+    assert hasRole2 => !lastReverted;
+}
+
+/// @rule: Once a sanction is added, it can be removed.
+/// The addSanction method can revert, if the sanctions count overflows.
+/// One has to assume the count is always < max_uint8.
+rule removedSanctionCanBeAdded(address account, uint16 CID) {
+    env e1;
+    env e2; require e2.msg.value == 0;
+    requireInvariant hasSanctionCountIsNonZero(e1, account, CID);
+    requireInvariant hasSanctionCountIsNonZero(e2, account, CID);
+
+    bool hasRole1 = hasRole(KYC_PROVIDER_ROLE(), e1.msg.sender);
+    bool hasRole2 = hasRole(KYC_PROVIDER_ROLE(), e2.msg.sender);
+    removeSanction(e1, account, CID);
+    assert hasRole1;
+
+    addSanction@withrevert(e2, account, CID);
+    assert hasRole2 => !lastReverted;
+}
+
+rule addingOrRemovingSanctionsAreIndependent(bool addOrRemove_A, bool addOrRemove_B) {
+    env eA;
+    env eB;
+    address accountA; uint16 CID_A;
+    address accountB; uint16 CID_B;
+    requireInvariant hasSanctionCountIsNonZero(eA, accountA, CID_A);
+    requireInvariant hasSanctionCountIsNonZero(eB, accountB, CID_B);
+    require (accountA == accountB && CID_A != CID_B) => (
+        (isSanctioned(eA, accountA, CID_A) && isSanctioned(eB, accountB, CID_B)) => getSanctionsCount(accountA) >= 2);
+
+    storage initState = lastStorage;
+    if(addOrRemove_A) {
+        addSanction(eA, accountA, CID_A) at initState;
+    }
+    else {
+        removeSanction(eA, accountA, CID_A) at initState;
+    }
+
+    if(addOrRemove_B) {
+        addSanction(eB, accountB, CID_B) at initState;
+    }
+    else {
+        removeSanction(eB, accountB, CID_B) at initState;
+    }
+    if(addOrRemove_A) {
+        addSanction@withrevert(eA, accountA, CID_A);
+    }
+    else {
+        removeSanction@withrevert(eA, accountA, CID_A);
+    }
+
+    assert !(accountA == accountB && CID_A == CID_B) => !lastReverted;
```

Comment:
very compilcated rule and if makes the rule inline 6 calls, better to split to two rules or have if parametric, when we have a `if (f.selector == ... )` the analysis can remove dead code but here the conditions or on a variables

Reply:
It's a common pattern when there are only two options. In the rule report it's easy to see which path was taken by the two booleans.